builtins: implement gateway_region

otan opened the pull request:
Implement the `gateway_region` builtin, which returns the region of the
current node.

Refs #57672

Release note (sql change): Implement the `gateway_region`, which returns
the region of the current node the connection is on.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58423)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Nit: Release notes might be better worded as "Release note (sql change): Implement the `gateway_region`, which returns the region of the connection's current node."

Code <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.  A few nits below.

> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58423#-:-MQDtxwe-PsL0SQajzJJ:b-bpd5fh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[docs/generated/sql/functions.md, line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58423#-MQDv8NsAAOW84r1FC_y:-MQDv8NsAAOW84r1FC_z:b-fztjk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a17b129e5d506effbaa5a1c68727abe0a9e3fbb/docs/generated/sql/functions.md#L870)):*
> ```Markdown
> <thead><tr><th>Function &rarr; Returns</th><th>Description</th></tr></thead>
> <tbody>
> <tr><td><a name="gateway_region"></a><code>gateway_region() &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns the region of the current node, as defined by the locality flag on startup.
> ```

Returns the region of the current node -> Returns the region of the connection's current node?

on startup -> on _node_ startup?
___
*[pkg/sql/logictest/testdata/logic_test/multiregion-invalid-locality, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58423#-MQE1wQl1PiZdHfHM8V8:-MQE1wQl1PiZdHfHM8V9:b-2wcfxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a17b129e5d506effbaa5a1c68727abe0a9e3fbb/pkg/sql/logictest/testdata/logic_test/multiregion-invalid-locality#L12)):*
> ```
> SELECT gateway_region()
> 
> query error no region set on the locality flag on this node
> ```

Copy pasta?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58423)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[docs/generated/sql/functions.md, line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58423#-MQDv8NsAAOW84r1FC_y:-MQEQPvH7yJGWnTVqfsy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a17b129e5d506effbaa5a1c68727abe0a9e3fbb/docs/generated/sql/functions.md#L870)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Returns the region of the current node -> Returns the region of the connection's current node?

on startup -> on _node_ startup?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/multiregion-invalid-locality, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58423#-MQE1wQl1PiZdHfHM8V8:-MQEQPUeB259zhWSpzs2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a17b129e5d506effbaa5a1c68727abe0a9e3fbb/pkg/sql/logictest/testdata/logic_test/multiregion-invalid-locality#L12)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Copy pasta?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: implement gateway_region

Implement the `gateway_region` builtin, which returns the region of the
current node.

Release note (sql change): Implement the gateway_region builtin, which
returns the region of the connection's current node.

otan commented:
bors r=ajstorm
